system-addon/content-src/components/TopSites/TopSite.jsx

```diff
@@ -175,24 +253,148 @@ export class TopSite extends React.PureComponent {
     </TopSiteLink>);
   }
 }
-TopSite.defaultProps = {link: {}};
+TopSite.defaultProps = {
+  link: {},
+  onDragStart() {}
+};
 
-export const TopSitePlaceholder = () => <TopSiteLink className="placeholder" />;
+export const TopSitePlaceholder = props => <TopSiteLink className="placeholder" isDraggable={false} {...props} />;
 
-export const TopSiteList = props => {
-  const topSites = props.TopSites.rows.slice(0, props.TopSitesCount);
-  const topSitesUI = [];
-  for (let i = 0, l = props.TopSitesCount; i < l; i++) {
-    const link = topSites[i];
-    topSitesUI.push(!link ? <TopSitePlaceholder key={i} /> : <TopSite
-      key={link.guid || link.url}
-      dispatch={props.dispatch}
-      link={link}
-      index={i}
-      intl={props.intl}
-      onEdit={props.onEdit} />);
-  }
-  return (<ul className="top-sites-list">
-    {topSitesUI}
-  </ul>);
-};
+export class TopSiteList extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = this.DEFAULT_STATE = {
+      draggedIndex: null,
+      draggedSite: null,
+      draggedTitle: null,
+      topSitesPreview: null
+    };
+    this.onDragStart = this.onDragStart.bind(this);
+    this.onDragEnd = this.onDragEnd.bind(this);
+    this.onDragEnter = this.onDragEnter.bind(this);
+    this.onDragLeave = this.onDragLeave.bind(this);
+    this.onDrop = this.onDrop.bind(this);
+  }
+  componentWillUpdate(nextProps) {
+    if (this.state.draggedSite) {
+      const prevTopSites = this.props.TopSites && this.props.TopSites.rows;
+      const newTopSites = nextProps.TopSites && nextProps.TopSites.rows;
+      if (prevTopSites && prevTopSites[this.state.draggedIndex] &&
+        prevTopSites[this.state.draggedIndex].url === this.state.draggedSite.url &&
+        (!newTopSites[this.state.draggedIndex] || newTopSites[this.state.draggedIndex].url !== this.state.draggedSite.url)) {
```

Comment:
This check kind of got out of hand. Can it be simpler maybe?

Reply:
I don't know if it breaks redux flat object something something. But if the .rows aren't the same Array object, maybe, it's different enough? Otherwise, maybe something like `rows.map(r => r.url).toString()` checking might be good enough to see if the urls are different?